### PR TITLE
relation: Add .readyset_explain

### LINKED ISSUE #44: Add `ActiveRecord::Relation#readyset_explain`
**Brief Description of Feature:**
We should add an easy way to invoke `EXPLAIN CREATE CACHE FROM ...` within the context of an ActiveRecord query via an `ActiveRecord::Relation#readyset_explain` method.

**Background:**
Users may want to get information about an ActiveRecord query like its ReadySet query ID, whether its supported by ReadySet, or ReadySet's rewritten version of the query. Adding an extension to `Relation` will make it easy for users to get this information.

**Detailed Explanation:**
Invoking `Relation#readyset_explain` will convert the query to SQL via `to_sql` and pass that SQL string to `EXPLAIN CREATE CACHE FROM <SQL string>` on ReadySet. The results should be returned either as a hash or as an object of a dedicated `Readyset::Explain` cache that lets users call methods to view the results.

---
## Optional information to add

**Proposed Solution:**
We should add this method onto the `RelationExtension` module being added in #27. The definition should look something like:
```ruby
def readyset_explain
  Readyset.explain(to_sql)
end
```
where `Readyset.explain` invokes `EXPLAIN CREATE CACHE FROM`.

### PR DESCRIPTION
**Pull request by ethan-readyset**
This commit adds a `.readyset_explain` method to our relation extension.
This method invokes `EXPLAIN CREATE CACHE` upstream on ReadySet and
returns information about the query from ReadySet.

Closes #44

### CODE REVIEW

File: spec/explain_spec.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+RSpec.describe Readyset::Explain do
+  describe '.call' do
+    it 'retrieves the explain information from ReadySet' do
+      explain = build(:explain)
+      raw_result = {
+        :'query id' => explain.id,
+        :'readyset supported' => explain.supported,
+        query: explain.text,
+      }
```
Comment (helpotters) [inline spec/explain_spec.rb:13]: This reflects what the `explain` factory is building. Provides visibility into what we're working with.

File: lib/readyset/explain.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    # Compares `self` with another `Explain` by comparing them attribute-wise.
+    #
+    # @param [Explain] other the `Explain` to which `self` should be compared
+    # @return [Boolean]
+    def ==(other)
+      id == other.id &&
+        text == other.text &&
+        supported == other.supported
+    end
```
Comment (helpotters) [inline lib/readyset/explain.rb:38]: Where did the motivation for the `#==` come from? It's a super neat inclusion.

File: spec/relation_extension_spec.rb [hunk 3]
```diff
@@ -29,9 +28,34 @@
...
+      explain = Readyset::Explain.new(id: 'q_0000000000000000',
+                                      text: 'SELECT * FROM "cats" WHERE ("id" = $1)',
+                                      supported: :yes)
+      allow(Readyset).to receive(:explain).with(query_string).and_return(explain)
+
+      output = query.readyset_explain
+
+      expect(output).to eq(explain)
+    end
+  end
```
Comment (helpotters) [inline spec/relation_extension_spec.rb:60]: This looks a bit similar to `ready_set_spec`, but nested one-level deeper. 

What was the use-case for `relation_extension` having its own interface for `explain`?

**Review by helpotters — COMMENTED** [review]
General feedback and a single question about `relation_extension`.

File: spec/relation_extension_spec.rb (hunk 3 above)
Comment (ethan-readyset) [reply spec/relation_extension_spec.rb:60]: The idea is that users can just call `.readyset_explain` on an arbitrary ActiveRecord expression like `User.where(active: true).readyset_explain`. Without this, users would need to do something like `Readyset.explain(User.where(active: true))`. It's not a huge difference, but I think the former is more ergonomic, and it more closely matches [the syntax for the regular `EXPLAIN` command](https://guides.rubyonrails.org/active_record_querying.html#running-explain)

File: lib/readyset/explain.rb (hunk 2 above)
Comment (ethan-readyset) [reply lib/readyset/explain.rb:38]: I added it so we can test for equality in tests

**Comment by ethan-readyset** [comment]
@helpotters Ready for another pass!

File: lib/readyset/explain.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
+module Readyset
+  class Explain
+    attr_reader :id, :text, :supported
+
+    # Gets information about the given query from ReadySet, including whether it's supported to be
+    # cached, its current status, the rewritten query text, and the query ID.
+    #
+    # The information about the given query is retrieved by invoking `EXPLAIN CREATE CACHE FROM` on
+    # ReadySet.
```
Comment (helpotters) [inline lib/readyset/explain.rb:9]: _**typo:**_ Module/Class responsibility descriptions should go above `class Explain`.

File: lib/readyset/caches.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+module Readyset
+  class Caches
```
Comment (helpotters) [inline lib/readyset/caches.rb:2]: _**todo:**_ We should add a description of the module's responsibility here

File: lib/readyset/caches.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    class << self
+      attr_reader :caches
+    end
+
+    def self.cache(id:, always: false)
```
Comment (helpotters) [inline lib/readyset/caches.rb:7]: **_suggestion:_** Does `always` as a boolean parameter work better as a separate method?

**Review by helpotters — CHANGES_REQUESTED** [review]
Let me know what you think! 

Resolve at your discretion, but aside from two minor typos and a possible refactor consideration, it's good to go.

File: lib/readyset/caches.rb (hunk 6 above)
Comment (ethan-readyset) [reply lib/readyset/caches.rb:9]: Hmm I'm not sure I follow -- can you give an example of what you mean?

This class defines a DSL where `.cache` is used to define caches to be used for cache migrations. `always` is a parameter for a given cache and indicates whether ReadySet will send queries to the upstream database if querying a cache fails for any reason. IMO, it's a _parameter_ of a cache as opposed to a separate entity, so it seemed appropriate to include it as a parameter of the `.cache` method

**Comment by ethan-readyset** [comment]
@helpotters Ready for another pass!

File: lib/readyset/caches.rb (hunk 6 above)
Comment (helpotters) [reply lib/readyset/caches.rb:9]: I guess it depends on the caller of `.cache`.

 Like if it knows it is using `always` or not, and if that conditional logic should be surfaced more explicitly to reflect the behavior of what the boolean parameter represents.

**Review by helpotters — APPROVED** [review]

**Pull request merged**